Review edits (#58)
dib-lab/2020-workflows-paper: content/04.workflows-and-software.md

The emergence and maturation of workflow systems designed with bioinformatic challenges in mind has revolutionized computing in data intensive biology [@doi:10.1038/s41592-018-0046-7].
Workflow systems contain powerful infrastructure for workflow management that can coordinate runtime behavior, self-monitor progress and resource usage, and compile reports documenting the results of a workflow (**Figure @fig:workflow**).

> These features ensure that the steps for data analysis are at least minimally documented and repeatable from start to finish.

Reviewer: "at least minimally documented and repeatable" - just to clarify, this is "completely repeatable" and "at least minimally documented"?

Author: yep, thanks!

> These features ensure that the steps for data analysis are at least minimally documented and repeatable from start to finish.

Author: ```suggestion
These features ensure that the steps for data analysis are repeatable and at least minimally documented from start to finish.
```

> These features ensure that the steps for data analysis are repeatable and at least minimally documented from start to finish.

Reviewer: maybe: minimally described? "documented" is such a loaded term.

> These features ensure that the steps for data analysis are repeatable and at least minimally documented from start to finish.

Author: ```suggestion
These features ensure that the steps for data analysis are repeatable and at least minimally described from start to finish.
```

In particular, research workflows require flexibility and assessment at every step: outliers and edge cases may reveal interesting biological differences, rather than sample processing or technical errors.
Many workflow systems can be used for either type, but we note cases where their properties facilitate one of these types over the other.

**Using workflows without learning workflow syntax**
While the benefits of executing an analysis within a data-centric workflow system are immense, the learning curve associated with command-line systems can be daunting.
It is possible to obtain the benefits of workflow systems without learning new syntax.
Websites like Galaxy, Cavatica, and EMBL-EBI MGnify offer online portals in which users build workflows around publicly-available or user-uploaded data [@doi:10.1093/nar/gky379; @doi:10.14694/EDBK_175029; @doi:10.1093/nar/gkz1035].
On the command line, many research groups have used workflow systems to build user-friendly pipelines that do not require learning or working with the underlying workflow software.

> These tools wrap one or many analysis steps (specified in an underlying workflow language) in a more user-friendly command-line tool that accepts user input and executes the analysis.

Reviewer: Maybe talk about pipelines instead of tools here? Matches language used in next sentence pretty well.

> These pipeline applications allow users to take advantage of the benefits of workflow software without needing to write the workflow syntax for each analysis step.

Reviewer: suggestion: remove "of the benefits of" => "allow users to take advantage of workflow software".

Could also add something installability here.

> These pipeline applications allow users to take advantage of the benefits of workflow software without needing to write the workflow syntax for each analysis step.

Author: ```suggestion
These pipeline applications allow users to take advantage of workflow software without needing to write the workflow syntax or manage software installation for each analysis step.
```

Some examples include the nf-core RNA-seq pipeline [@url:https://github.com/nf-core/rnaseq/; @doi:10.1038/s41587-020-0439-x], the PiGx genomic analysis toolkit [@doi:10.1093/gigascience/giy123], the ATLAS metagenome assembly and binning pipeline [@url:https://github.com/metagenome-atlas/atlas; @doi:10.1101/737528], the Sunbeam metagenome analysis pipeline [@url:https://github.com/sunbeam-labs/sunbeam; @doi:10.1186/s40168-019-0658-x], and two from our own lab, the dammit eukaryotic transcriptome annotation pipeline [@url:https://github.com/dib-lab/dammit] and the elvers *de novo* transcriptome pipeline [@url:https://github.com/dib-lab/elvers].

> These tools allow users to take advantage of the benefits of workflow software without needing to invest in curating and writing their own pipeline. Rather than writing the workflow syntax for each analysis step, the user can specify data and parameters in a configuration file to customize the run. The majority of these workflows are production-level workflows designed to execute a series of standard steps, but many provide varying degrees of customizability ranging from tool choice to parameter specification.

Reviewer: Maybe: "these pipeline applications", or "these customizable pipelines", instead of "these tools"?

**Choosing a workflow system**
If your use case extends beyond these tools, there are several scriptable workflow systems that offer comparable benefits for carrying out your own data-intensive analyses.
Each has it own strengths, meaning each workflow software will meet an individuals computing goals differently (see **Table @tbl:workflows**).

> Our lab has adopted Snakemake [@doi:10.1093/bioinformatics/bts480 @doi:10.5281/zenodo.4067137], in part due to its similarity and integration with Python, its flexibility for building and testing new analyses in different languages, and its intuitive integration with software management tools (described below).

Reviewer: suggestion: remove "similarity and" and just leave it as "in part due to its integration with Python,"

> Our lab has adopted Snakemake [@doi:10.1093/bioinformatics/bts480 @doi:10.5281/zenodo.4067137], in part due to its similarity and integration with Python, its flexibility for building and testing new analyses in different languages, and its intuitive integration with software management tools (described below).

Author: ```suggestion
Our lab has adopted Snakemake [@doi:10.1093/bioinformatics/bts480 @doi:10.5281/zenodo.4067137], in part due to its integration with Python, its flexibility for building and testing new analyses in different languages, and its intuitive integration with software management tools (described below).
```

Snakemake and Nextflow [@doi:10.1038/nbt.3820] are commonly used for developing new research pipelines, where flexibility and iterative, branching development is a key feature. Common Workflow Language (CWL) and Workflow Description Language (WDL) are workflow specification formats that are more geared towards scalability, making them ideal for production-level pipelines with hundreds of thousands of samples [@doi:10.6084/m9.figshare.3115156.v2]. WDL and CWL are commonly executed on platforms such as Terra [@url:https://terra.bio] or Seven Bridges Platform [@url:https://www.sevenbridges.com/platform/].

> Language-specific workflow systems, such as ROpenSci's Drake [@doi:10.21105/joss.00550], are more limited in the scope of tasks they can execute, but are powerful within their language and easier to integrate for those comfortable with that language.

Author: ```suggestion
Language-specific workflow systems, such as ROpenSci's Drake [@doi:10.21105/joss.00550], can take full advantage of the language's internal data structures and provide automation and reproducibility benefits for workflows executed primarily within the language ecosystem. 
```

> Language-specific workflow systems, such as ROpenSci's Drake [@doi:10.21105/joss.00550], can take full advantage of the language's internal data structures and provide automation and reproducibility benefits for workflows executed primarily within the language ecosystem. 

Reviewer: suggestion: add comma after "data structures"

> Language-specific workflow systems, such as ROpenSci's Drake [@doi:10.21105/joss.00550], can take full advantage of the language's internal data structures and provide automation and reproducibility benefits for workflows executed primarily within the language ecosystem. 

Author: ```suggestion
Language-specific workflow systems, such as ROpenSci's Drake [@doi:10.21105/joss.00550], can take full advantage of the language's internal data structures, and provide automation and reproducibility benefits for workflows executed primarily within the language ecosystem. 
```

> Language-specific workflow systems, such as ROpenSci's Drake [@doi:10.21105/joss.00550], are more limited in the scope of tasks they can execute, but are powerful within their language and easier to integrate for those comfortable with that language.

Reviewer: instead of "are powerful within their language" say "can take full advantage of the language's internal data structures"? (This is a soft suggestion, doesn't quite capture it :)

These tools are heterogeneous in nature: they are written by researchers working in different coding languages, with varied approaches to software design and optimization, and often for specific analysis goals.
Each program has a number of other programs it depends upon to function ("dependencies"), and as software changes over time to meet research needs, the results may change, even when run with identical parameters.
As a result, it is critical to take an organized approach to installing, managing, and keeping track of software and software versions.
On many compute systems, system-wide software management is overseen by system administrators, who ensure commonly-used and requested software is installed into a "module" system available to all users.
Unfortunately, this system limits software version transparency and does not lend itself well to exploring new workflows and software, as researchers do not have permission to install software themselves.

> To meet this need, most workflow managers integrate with software management systems that handle software installation, management, and packaging, alleviating problems that arise from dependencies and facilitating documentation of software versions.

Reviewer: ...that arise from _complex_ dependencies, maybe.

> To meet this need, most workflow managers integrate with software management systems that handle software installation, management, and packaging, alleviating problems that arise from dependencies and facilitating documentation of software versions.

Author: ```suggestion
To meet this need, most workflow managers integrate with software management systems that handle software installation, management, and packaging, alleviating problems that arise from complex dependencies and facilitating documentation of software versions.
```

Heavyweight software management systems package not only the software of interest, but also the runtime environment information, with the goal of ensuring perfect reproducibility in software installation over time.
Tools such as singularity and docker [@doi:10.1016/j.cels.2018.03.014; @doi:10.1038/s41592-018-0046-7; @doi:10.1371/journal.pone.0177459; @url:https://dl.acm.org/doi/10.5555/2600239.2600241] wrap software environments in "containers" that capture and reproduce the runtime environment information.
Container-based management is particularly useful for systems where some dependencies may not be installable by lightweight managers. However, software installation within these containers can be limited by similar reproducibility issues, including changes in dependency installations over time.
"Functional package managers" such as GNU Guix and Nix strictly require all dependency and configuration details be encoded within each software package, providing the most comprehensively reproducible installations.

> These have begun to be integrated into some bioinformatic tools, but have a steeper learning curve for independent use [@doi:10.1093/gigascience/giy123].

Reviewer: question: I'm pretty sure that docker and singularity require that they be installed/permitted by the sysadmins, while conda does not. what about Guix and Nix? I know it's one of the big reasons why I adopted conda. @luizirber do you know?

Author: This seems to be true based on my limited knowledge. @luizirber?

> These have begun to be integrated into some bioinformatic tools, but have a steeper learning curve for independent use [@doi:10.1093/gigascience/giy123].

Author: ```suggestion
These have begun to be integrated into some bioinformatic tools [@doi:10.1093/gigascience/giy123], but have a steeper learning curve for independent use.
In addition, standard installation of these managers requires system-wide installation permissions, requiring assistance from system administrators on most high-performance computing systems.
```

Author: @luizirber @camillescott accurate / ok wording?

Reviewer: it's correct, I think. But "requiring assistance" doesn't fully capture the horror of the situation, which involves policy level support for such things. I'm not sure it's worth iterating on - or at least I can't come up with a better phrasing ;) - but for me, it's definitely a blocker when root needs to do something!

Author: i know, i just can't think of a better way to put it!

![**The conda package and environment manager simplifies software installation and management.**

> **A. Conda Recipe Repositories:** Each program distributed via Conda has a "recipe" describing all software dependencies needed for installation using Conda (each of which must also be installable via Conda). Recipes are stored and managed in the cloud in separate "channels", some of which specialize in particular fields or languages (e.g. the "bioconda" channel specializes in bioinformatic software) [@doi:10.1038/s41592-018-0046-7]. **B. Use Conda Environments to Avoid Installation Conflicts:**  Conda does not require root

Author: ```suggestion
**A. Conda Recipe Repositories:** Each program distributed via Conda has a "recipe" describing all software dependencies needed for installation using Conda (each of which must also be installable via Conda). Recipes are stored and managed in the cloud in separate "channels", some of which specialize in particular fields or languages (e.g. the "bioconda" channel specializes in bioinformatic software, while the "conda-forge" channel is a more general effort to provide and maintain standardized conda packages for a wide range of software) [@doi:10.1038/s41592-018-0046-7]. **B. Use Conda Environments to Avoid Installation Conflicts:**  Conda does not require root
```

privileges for software installation, thus enabling use by researchers working on shared cluster systems. However, even user-based software installation can encounter dependency conflicts. For example, you might need to use python2 to install and run a program (e.g. older scripts written by members of your lab), while also using snakemake to execute your workflows (requires python>=3.5). By installing each program into an isolated "environment" that contains only the software required to run that

> program, you can ensure all programs used throughout your analysis will run without issue. Using small, separate environments for your software and building many simple environments to accommodate different steps in your workflow is critical for reducing the amount of time it takes conda to resolve dependency conflicts between different software tools ("solve" an environment). Conda virtual environments can be created and installed either on the command line, or via an environment YAML file, as shown. In this case, the environment file

Author: ```suggestion
program, you can ensure all programs used throughout your analysis will run without issue. Using small, separate environments for your software, specifying the desired software version, and building many simple environments to accommodate different steps in your workflow is critical for reducing the amount of time it takes conda to resolve dependency conflicts between different software tools ("solve" an environment). Conda virtual environments can be created and installed either on the command line, or via an environment YAML file, as shown. In this case, the environment file
```

> also specifies which conda channels to search and download programs from. When specified in a YAML file, conda environments are easily transferable between computers and operating systems. Further, because the version of each package installed in an environment is recorded, workflow reproducibility is enhanced. While Conda may be superseded by alternative solutions [@url:https://github.com/QuantStack/mamba], this model of software installation and management will likely

Author: ```suggestion
also specifies which conda channels to search and download programs from. 
When specified in a YAML file, conda environments are easily transferable between computers and operating systems. 
Broad community adoption has resulted in a proliferation of both conda-installable scientific software and tools that leverage conda installation specifications.
For example, the Mamba package manager is an open source reimplementation of the conda manager that can install conda-style environments with increased efficiency  [@url:https://github.com/QuantStack/mamba]. 
The BioContainers Registry is a project that automatically builds and distributes docker and singularity containers for bioinformatics software packages using each package's conda installation recipe [@doi:https://doi.org/10.1101/2020.07.21.187609]. 
```

> persist.](images/conda_figure_ABC.svg){#fig:conda_figure_ABC height=7in}

Author: ```suggestion
](images/conda_figure_ABC.svg){#fig:conda_figure_ABC height=7in}
```

For production workflows that require maximal reproducibility, it is worth the larger investment required to use heavyweight systems.
This is particularly true for advanced users who can more easily navigate the steps required for utilizing these tools.
Container-based software installation via docker and singularity are common for production-level workflows, and Guix and Nix-based solutions are gaining traction.
Importantly, the needs and constraints of a project can evolve over time, as may the system of choice.

**Integrating software management within workflows**
Workflow systems provide seamless integration with a number of software management tools.
Each workflow system requires different specification for initiation of software management, but typically requires about one additional line of code per step that requires the use of software.
If the software management tool is installed locally, the workflow will automatically download and install the specified environment or container and use it for specified step.

> In our experience, the complete solution for using scientific software involves a combination of approaches. Interactive and exploratory analyses conducted in IDEs and jupyter notebooks (usually with local software installation with conda) are useful for developing an analysis strategy and creating an initial workflow. This is then followed by workflow-integrated software management via conda, singularity, or nixOS for executing the resulting workflow on many samples. This process is iterative as we build and extend our analyses.

Reviewer: Re "This process is iterative as we build and extend our analyses" - grammatically questionable :). Maybe "We then iterate as we build and extend our analyses." Although come to think of it I'm not sure what we're iterating on. We then iteratively extend the workflow?

> In our experience, the complete solution for using scientific software involves a combination of approaches. Interactive and exploratory analyses conducted in IDEs and jupyter notebooks (usually with local software installation with conda) are useful for developing an analysis strategy and creating an initial workflow. This is then followed by workflow-integrated software management via conda, singularity, or nixOS for executing the resulting workflow on many samples. This process is iterative as we build and extend our analyses.

Author: ```suggestion
In our experience, the complete solution for using scientific software involves a combination of approaches. Interactive and exploratory analyses conducted in IDEs and jupyter notebooks (usually with local software installation with conda) are useful for developing an analysis strategy and creating an initial workflow. This is then followed by workflow-integrated software management via conda, singularity, or nixOS for executing the resulting workflow on many samples. This process not linear: we often cycle between exploratory testing and automation as we iteratively extend our analyses.
```

These tools are heterogeneous in nature: they are written by researchers working in different coding languages, with varied approaches to software design and optimization, and often for specific analysis goals.
Each program has a number of other programs it depends upon to function ("dependencies"), and as software changes over time to meet research needs, the results may change, even when run with identical parameters.
As a result, it is critical to take an organized approach to installing, managing, and keeping track of software and software versions.
On many compute systems, system-wide software management is overseen by system administrators, who ensure commonly-used and requested software is installed into a "module" system available to all users.

> Unfortunately, this system limits reproducibility and does not lend itself well to exploring new workflows and software, as researchers do not have permission to install software themselves.

Reviewer 2: This sentence doesn't make it clear *why* software installed in modules limits reproducibility. Is it by shielding the user from version updates or something?

Author: I suggested a change below to address this

> Unfortunately, this system limits reproducibility and does not lend itself well to exploring new workflows and software, as researchers do not have permission to install software themselves.

Author: ```suggestion
Unfortunately, this system limits software version transparency and does not lend itself well to exploring new workflows and software, as researchers do not have permission to install software themselves.
```

Conda handles both cluster permission and version conflict issues with a user-based software environment system, and features a straightforward "recipe" system which simplifies the process of making new software installable (including simple management of versions and updates).
These features have led to widespread adoption within the bioinformatic community: packages for new software become quickly available, and can be installed easily across platforms.
However, conda does not attempt to isolate software installations or standardize the runtime environment, meaning installations will not be completely reproducible over time.
Heavyweight software management systems package not only the software of interest, but also the runtime environment information, with the goal of ensuring perfect reproducibility in software installation over time.
Tools such as singularity and docker [@doi:10.1016/j.cels.2018.03.014; @doi:10.1038/s41592-018-0046-7; @doi:10.1371/journal.pone.0177459; @url:https://dl.acm.org/doi/10.5555/2600239.2600241] wrap software environments in "containers" that capture and reproduce the runtime environment information.

> However, software installation within these containers can be limited by similar issues, including changes in dependency installations over time.

Author: ```suggestion
Container-based management is particularly useful for systems where some dependencies may not be installable by lightweight managers. However, software installation within these containers can be limited by similar reproducibility issues, including changes in dependency installations over time.
```

> [@doi:10.1093/gigascience/giy123].

Author: ```suggestion
```

"Functional package managers" such as GNU Guix and Nix strictly require all dependency and configuration details be encoded within each software package, providing the most comprehensively reproducible installations.
These have begun to be integrated into some bioinformatic tools, but have a steeper learning curve for independent use [@doi:10.1093/gigascience/giy123].

![**The conda package and environment manager simplifies software installation and management.**

> **A. Conda Recipe Repositories:** Each program distributed via Conda has a "recipe" describing all software dependencies needed for installation using Conda (each of which must also be installable via Conda). Recipes are stored and managed in the cloud in separate "channels", some of which specialize in particular fields or languages (e.g. the "bioconda" channel specializes in bioinformatic software, the "r" channel specializes in R language packages) [@doi:10.1038/s41592-018-0046-7]. **B. Use Conda Environments to Avoid Installation Conflicts:**  Conda does not require root

Reviewer 2: ```suggestion
**A. Conda Recipe Repositories:** Each program distributed via Conda has a "recipe" describing all software dependencies needed for installation using Conda (each of which must also be installable via Conda). Recipes are stored and managed in the cloud in separate "channels", some of which specialize in particular fields or languages (e.g. the "bioconda" channel specializes in bioinformatic software) [@doi:10.1038/s41592-018-0046-7]. **B. Use Conda Environments to Avoid Installation Conflicts:**  Conda does not require root
```

Reviewer 2: Most R packages are now installed through `conda-forge` instead of through the `r` channel. I don't think we should get into that, but I think we should remove the `r` channel reference as its misleading to actual practices.
@luizirber do you have any thoughts about this?

Author: perhaps we should add something about conda-forge instead?

Reviewer: and/or the broad community adoption of conda could be mentioned somewhere. It's truly amazing.

Might also mention that biocontainers automatically builds containers from bioconda

privileges for software installation, thus enabling use by researchers working on shared cluster systems. However, even user-based software installation can encounter dependency conflicts. For example, you might need to use python2 to install and run a program (e.g. older scripts written by members of your lab), while also using snakemake to execute your workflows (requires python>=3.5). By installing each program into an isolated "environment" that contains only the software required to run that

> program, you can ensure all programs used throughout your analysis will run without issue. Using small, separate environments for your software and building many simple environments to accommodate different steps in your workflow also reduces the amount of time it takes conda to resolve dependency conflicts between different software tools ("solve" an environment). Conda virtual environments can be created and installed either on the command line, or via an environment YAML file, as shown. In this case, the environment file

Author: ```suggestion
program, you can ensure all programs used throughout your analysis will run without issue. Using small, separate environments for your software and building many simple environments to accommodate different steps in your workflow is critical for reducing the amount of time it takes conda to resolve dependency conflicts between different software tools ("solve" an environment). Conda virtual environments can be created and installed either on the command line, or via an environment YAML file, as shown. In this case, the environment file
```

For production workflows that require maximal reproducibility, it is worth the larger investment required to use heavyweight systems.
This is particularly true for advanced users who can more easily navigate the steps required for utilizing these tools.
Container-based software installation via docker and singularity are common for production-level workflows, and Guix and Nix-based solutions are gaining traction.
Importantly, the needs and constraints of a project can evolve over time, as may the system of choice.

**Integrating software management within workflows**
Workflow systems provide seamless integration with a number of software management tools.
Each workflow system requires different specification for initiation of software management, but typically requires about one additional line of code per step that requires the use of software.
If the software management tool is installed locally, the workflow will automatically download and install the specified environment or container and use it for specified step.

> In our experience, the complete solution for using scientific software involves a combination of approaches. Interactive and exploratory analyses conducted in IDEs and jupyter notebooks (usually with local software installation with conda) is useful for developing an analysis strategy and creating an initial workflow. This is then followed by workflow-integrated software management via conda, singularity, or nixOS for executing the resulting workflow on many samples. This process is iterative as we build and extend our analyses.

Author: ```suggestion
In our experience, the complete solution for using scientific software involves a combination of approaches. Interactive and exploratory analyses conducted in IDEs and jupyter notebooks (usually with local software installation with conda) are useful for developing an analysis strategy and creating an initial workflow. This is then followed by workflow-integrated software management via conda, singularity, or nixOS for executing the resulting workflow on many samples. This process is iterative as we build and extend our analyses.
```

Unfortunately, this system limits software version transparency and does not lend itself well to exploring new workflows and software, as researchers do not have permission to install software themselves.
To meet this need, most workflow managers integrate with software management systems that handle software installation, management, and packaging, alleviating problems that arise from complex dependencies and facilitating documentation of software versions.
Software management systems range from lightweight systems that manage only the software and its dependencies, to heavyweight systems that control for all aspects of the runtime and operating system, ensuring 100% reproducibility of results across computational platforms and time.

On the lightweight end, the conda package manager has emerged as a leading software management solution for research workflows (**Figure @fig:conda_figure_ABC**).
Conda handles both cluster permission and version conflict issues with a user-based software environment system, and features a straightforward "recipe" system which simplifies the process of making new software installable (including simple management of versions and updates).

> These features have led to widespread adoption within the bioinformatic community: packages for new software become quickly available, and can be installed easily across platforms.

Reviewer: suggestion: "bioinformatic community" => "bioinformatics community"

> These features have led to widespread adoption within the bioinformatic community: packages for new software become quickly available, and can be installed easily across platforms.

Author: ```suggestion
These features have led to widespread adoption within the bioinformatics community: packages for new software become quickly available, and can be installed easily across platforms.
```

> However, conda does not attempt to isolate software installations or standardize the runtime environment, meaning installations will not be completely reproducible over time.

Reviewer: hmm, this:
>However, conda does not attempt to isolate software installations or standardize the runtime environment, meaning installations will not be completely reproducible over time.

doesn't quite make sense to me. Conda _does_ isolate software installs, just not perfectly. And I don't know what "standardize the runtime environment" means, although I can guess.

Maybe:
> However, conda does not completely isolate software installations and aims neither for bitwise reproducibility nor long-term archiving of install packages, meaning installations will not be completely reproducible over time.

Author: yes, much better -thanks!

> However, conda does not attempt to isolate software installations or standardize the runtime environment, meaning installations will not be completely reproducible over time.

Author: ```suggestion
However, conda does not completely isolate software installations and aims neither for bitwise reproducibility nor long-term archiving of install packages, meaning installations will not be completely reproducible over time.
```
